File: Polars.FSharp/IO/JSON/Read.fs
namespace Polars.FSharp

open Polars.NET.Core

[<AutoOpen>]
module JsonRead =
    open System.IO
    type LazyFrame with
        /// <summary>
        /// Lazily read a NDJSON file.
        /// </summary>
        static member ScanNdjson(path: string,
                                ?schema: PolarsSchema,
                                ?inferSchemaLen: uint64,
                                ?batchSize: uint64,
                                ?nRows: uint64,
                                ?lowMemory: bool,
                                ?rechunk: bool,
                                ?ignoreErrors: bool,
                                ?rowIndexName: string,
                                ?rowIndexOffset: uint32,
                                ?includePathColumn: string,
                                ?cloudOptions: CloudOptions) : LazyFrame =
            
            let schemaHandle = match schema with Some s -> s.Handle | None -> null
            let inferLen = Option.toNullable inferSchemaLen
            let batch = Option.toNullable batchSize
            let rows = Option.toNullable nRows
            
            let lowMem = defaultArg lowMemory false
            let rechk = defaultArg rechunk false
            let ignoreErr = defaultArg ignoreErrors false
            
            let idxName = Option.toObj rowIndexName
            let idxOffset = defaultArg rowIndexOffset 0u
            let pathCol = Option.toObj includePathColumn

            // Parse Cloud Options
            let cProv, cRet, cToMs, cInitMs, cMaxMs, cCache, cKeys, cVals = 
                CloudOptions.ParseCloudOptions cloudOptions

            let h = PolarsWrapper.ScanNdjson(
                path, 
                schemaHandle, 
                batch, 
                inferLen, 
                rows, 
                lowMem, 
                rechk, 
                ignoreErr, 
                idxName, 
                idxOffset, 
                pathCol,
                // Cloud
                cProv,
                cRet,
                cToMs,
                cInitMs,
                cMaxMs,
                cCache,
                cKeys,
                cVals
            )
            new LazyFrame(h)

        /// <summary>
        /// Lazily read NDJSON from in-memory bytes.
        /// </summary>
        static member ScanNdjson(buffer: byte[],
                                ?schema: PolarsSchema,
                                ?inferSchemaLen: uint64,
                                ?batchSize: uint64,
                                ?nRows: uint64,
                                ?lowMemory: bool,
                                ?rechunk: bool,
                                ?ignoreErrors: bool,
                                ?rowIndexName: string,
                                ?rowIndexOffset: uint32) : LazyFrame =
            
            let schemaHandle = match schema with Some s -> s.Handle | None -> null
            let inferLen = Option.toNullable inferSchemaLen
            let batch = Option.toNullable batchSize
            let rows = Option.toNullable nRows
            
            let lowMem = defaultArg lowMemory false
            let rechk = defaultArg rechunk false
            let ignoreErr = defaultArg ignoreErrors false
            
            let idxName = Option.toObj rowIndexName
            let idxOffset = defaultArg rowIndexOffset 0u
            
            let h = PolarsWrapper.ScanNdjson(
                buffer, 
                schemaHandle, 
                batch, 
                inferLen, 
                rows, 
                lowMem, 
                rechk, 
                ignoreErr, 
                idxName, 
                idxOffset, 
                null 
            )
            new LazyFrame(h)

        /// <summary>
        /// Lazily read NDJSON from a Stream.
        /// </summary>
        static member ScanNdjson(stream: Stream,
                                ?schema: PolarsSchema,
                                ?inferSchemaLen: uint64,
                                ?batchSize: uint64,
                                ?nRows: uint64,
                                ?lowMemory: bool,
                                ?rechunk: bool,
                                ?ignoreErrors: bool,
                                ?rowIndexName: string,
                                ?rowIndexOffset: uint32) : LazyFrame =
            
            use ms = new MemoryStream()
            stream.CopyTo ms
            let bytes = ms.ToArray()

            LazyFrame.ScanNdjson(
                bytes,
                ?schema=schema,
                ?inferSchemaLen=inferSchemaLen,
                ?batchSize=batchSize,
                ?nRows=nRows,
                ?lowMemory=lowMemory,
                ?rechunk=rechunk,
                ?ignoreErrors=ignoreErrors,
                ?rowIndexName=rowIndexName,
                ?rowIndexOffset=rowIndexOffset
            )
    type DataFrame with
        /// <summary>
        /// Read a JSON file into a DataFrame.
        /// </summary>
        static member ReadJson(path: string, 
                            ?columns: string seq, 
                            ?schema: PolarsSchema, 
                            ?inferSchemaLen: uint64, 
                            ?batchSize: uint64, 
                            ?ignoreErrors: bool,
                            ?jsonFormat: JsonFormat) : DataFrame =
            
            let cols = columns |> Option.map Seq.toArray |> Option.toObj
            
            let schemaHandle = match schema with Some s -> s.Handle | None -> null
            
            let inferLen = Option.toNullable inferSchemaLen
            let batch = Option.toNullable batchSize
            
            let ignoreErr = defaultArg ignoreErrors false
            
            let fmt = defaultArg jsonFormat JsonFormat.Json

            let h = PolarsWrapper.ReadJson(path, cols, schemaHandle, inferLen, batch, ignoreErr, fmt.ToNative())
            new DataFrame(h)

        /// <summary>
        /// Read JSON from in-memory bytes.
        /// </summary>
        static member ReadJson(buffer: byte[], 
                            ?columns: string seq, 
                            ?schema: PolarsSchema, 
                            ?inferSchemaLen: uint64, 
                            ?batchSize: uint64, 
                            ?ignoreErrors: bool,
                            ?jsonFormat: JsonFormat) : DataFrame =
            
            let cols = columns |> Option.map Seq.toArray |> Option.toObj
            let schemaHandle = match schema with Some s -> s.Handle | None -> null
            let inferLen = Option.toNullable inferSchemaLen
            let batch = Option.toNullable batchSize
            let ignoreErr = defaultArg ignoreErrors false
            
            let fmt = defaultArg jsonFormat JsonFormat.Json

            let h = PolarsWrapper.ReadJson(buffer, cols, schemaHandle, inferLen, batch, ignoreErr, fmt.ToNative())
            new DataFrame(h)

        /// <summary>
        /// Read JSON from a Stream.
        /// </summary>
        static member ReadJson(stream: Stream, 
                            ?columns: string seq, 
                            ?schema: PolarsSchema, 
                            ?inferSchemaLen: uint64, 
                            ?batchSize: uint64, 
                            ?ignoreErrors: bool,
                            ?jsonFormat: JsonFormat) : DataFrame =
            
            use ms = new MemoryStream()
            stream.CopyTo ms
            let bytes = ms.ToArray()
            
            DataFrame.ReadJson(
                bytes, 
                ?columns=columns, 
                ?schema=schema, 
                ?inferSchemaLen=inferSchemaLen, 
                ?batchSize=batchSize, 
                ?ignoreErrors=ignoreErrors, 
                ?jsonFormat=jsonFormat
            )

File: Polars.CSharp/IO/JSON/Read.cs
using Polars.NET.Core;

namespace Polars.CSharp;
public partial class DataFrame : IDisposable,IEnumerable<Series>,IPolarsDataFrame
{
    /// <summary>
    /// Read a JSON file into a DataFrame.
    /// </summary>
    /// <param name="path">Path to the JSON file.</param>
    /// <param name="columns">Select specific columns to read.</param>
    /// <param name="schema">Manually specify the schema (recommended for stability).</param>
    /// <param name="inferSchemaLen">Number of rows to scan for schema inference (default: scan all).</param>
    /// <param name="batchSize">Batch size for reading (optimization).</param>
    /// <param name="ignoreErrors">Ignore parsing errors (skip malformed lines).</param>
    /// <param name="jsonFormat">Format: Json Array or Json Lines (NDJSON).</param>
    public static DataFrame ReadJson(
        string path,
        string[]? columns = null,
        IntoSchema? schema = null,
        ulong? inferSchemaLen = null,
        ulong? batchSize = null,
        bool ignoreErrors = false,
        JsonFormat jsonFormat = JsonFormat.Json)
    {
        if (!File.Exists(path)) 
            throw new FileNotFoundException($"JSON file not found: {path}");

        var schemaHandle = schema?.Consume().Handle;

        var h = PolarsWrapper.ReadJson(
            path,
            columns,
            schemaHandle,
            inferSchemaLen,
            batchSize,
            ignoreErrors,
            jsonFormat.ToNative()
        );

        return new DataFrame(h);
    }

    // ---------------------------------------------------------
    // Read JSON (Memory / Bytes)
    // ---------------------------------------------------------

    /// <summary>
    /// Read JSON from an in-memory byte array.
    /// </summary>
    public static DataFrame ReadJson(
        byte[] buffer,
        string[]? columns = null,
        IntoSchema? schema = null,
        ulong? inferSchemaLen = null,
        ulong? batchSize = null,
        bool ignoreErrors = false,
        JsonFormat jsonFormat = JsonFormat.Json)
    {
        var schemaHandle = schema?.Consume().Handle;

        var h = PolarsWrapper.ReadJson(
            buffer,
            columns,
            schemaHandle,
            inferSchemaLen,
            batchSize,
            ignoreErrors,
            jsonFormat.ToNative()
        );

        return new DataFrame(h);
    }

    // ---------------------------------------------------------
    // Read JSON (Stream)
    // ---------------------------------------------------------

    /// <summary>
    /// Read JSON from a Stream.
    /// </summary>
    public static DataFrame ReadJson(
        Stream stream,
        string[]? columns = null,
        IntoSchema? schema = null,
        ulong? inferSchemaLen = null,
        ulong? batchSize = null,
        bool ignoreErrors = false,
        JsonFormat jsonFormat = JsonFormat.Json)
    {
        using var ms = new MemoryStream();
        stream.CopyTo(ms);
        
        return ReadJson(
            ms.ToArray(), 
            columns, 
            schema, 
            inferSchemaLen, 
            batchSize, 
            ignoreErrors, 
            jsonFormat
        );
    }
}